Source/AGXUnreal/Private/Terrain/AGX_MovableTerrainComponent.cpp

```diff
@@ -0,0 +1,923 @@
+#include "Terrain/AGX_MovableTerrainComponent.h"
+#include "Terrain/AGX_ShovelComponent.h"
+#include "Shapes/HeightFieldShapeBarrier.h"
+#include "Shapes/AGX_ShapeComponent.h"
+#include "Materials/AGX_ShapeMaterial.h"
+#include "Materials/AGX_TerrainMaterial.h"
+#include "Utilities/AGX_StringUtilities.h"
+#include "Utilities/AGX_NotificationUtilities.h"
+#include "AGX_InternalDelegateAccessor.h"
+#include "AGX_RigidBodyComponent.h"
+#include "AGX_LogCategory.h"
+#include "AGX_Simulation.h"
+#include "TimerManager.h"
+#include "Engine/SimpleConstructionScript.h"
+#include "Engine/SCS_Node.h"
+#include "NiagaraComponent.h"
+#include "NiagaraDataInterfaceArrayFunctionLibrary.h"
+#include "NiagaraFunctionLibrary.h"
+
+UAGX_MovableTerrainComponent::UAGX_MovableTerrainComponent(
+	const FObjectInitializer& ObjectInitializer)
+	: UProceduralMeshComponent(ObjectInitializer)
+{
+	PrimaryComponentTick.bCanEverTick = true;
+	SetCanEverAffectNavigation(false);
+}
+
+void UAGX_MovableTerrainComponent::BeginPlay()
+{
+	Super::BeginPlay();
+
+	// Create Heights
+	InitializeHeights();
+
+	// Create Native
+	CreateNative();
+	
+	// Create Mesh(s) and Tile(s)
+	InitializeMesh();
+
+	// Update Mesh each tick
+	if (UAGX_Simulation* Simulation = UAGX_Simulation::GetFrom(this))
+	{
+		PostStepForwardHandle =
+			FAGX_InternalDelegateAccessor::GetOnPostStepForwardInternal(*Simulation)
+				.AddLambda(
+					[this](double)
+					{
+						//Update CurrentHeights
+						NativeBarrier.GetHeights(CurrentHeights, true);
+
+						//Rebuild mesh if needed
+						auto DirtyHeights = NativeBarrier.GetModifiedVertices();
+						if (DirtyHeights.Num() > 0)
+							UpdateMesh(DirtyHeights);
+					});
+	}
+
+	// Init particles
+	if (bEnableParticleRendering)
+	{
+		if (!InitializeParticles())
+		{
+			UE_LOG(
+				LogAGX, Warning,
+				TEXT("InitializeParticles returned false in AGX_MovableTerrainComponent '%s'. "
+					 "Ensure the selected Niagara System is valid."),
+				*GetName());
+		}
+	}
+
+	// Unreal Collision
+	this->SetCollisionEnabled(AdditionalUnrealCollision);
+}
+
+void UAGX_MovableTerrainComponent::CreateNative()
+{
+	// Make sure OwningRigidBody is created, if there is one
+	UAGX_RigidBodyComponent* OwningRigidBody =
+		FAGX_ObjectUtilities::FindFirstAncestorOfType<UAGX_RigidBodyComponent>(*this);
+	if (OwningRigidBody)
+		OwningRigidBody->GetOrCreateNative();
+
+	// Resolution
+	FIntVector2 TerrainResolution = GetTerrainResolution();
+
+	// Allocate Native
+	NativeBarrier.AllocateNative(
+		TerrainResolution.X, TerrainResolution.Y, ElementSize, CurrentHeights, BedHeights);
+
+	// Attach to RigidBody
+	if (OwningRigidBody)
+		OwningRigidBody->GetNative()->AddTerrain(&NativeBarrier);
```

Comment:
I suggest simply exposing a function in the TerrainBarrier similar to
`
```c++
FHeightFieldShapeBarrier GetHeightField() const
{
    check(HasNative());
    return FHeightFieldShapeBarrier(NativeRef->Native->getHeightField());
} 
```

Or whatever the API calls are in AGX (I have not looked this up). 

Then in your MovableTerrain::CreateNative function, you can do 

```
auto HeightField = NativeBarrier.GetHeightField();
OwningRigidBody->GetNative()->AddShape(HeightField);
```

Reply:
![image](https://github.com/user-attachments/assets/fbc1f3bd-6896-4885-83d8-36d516702d36)

That std::move in FHeightFieldShapeBarrier's constructor makes me suspicious... If I use this, will the TerrainBarrie loose its ownership over the native shape/geometry?

Reply:
I see why it can seem so, but the unique_ptr only uniquely owns a `FGeometryAndShapeRef` which in turn has a `agxCollide::ShapeRef` which is reference counted and the underlying `agxCollide::Shape` object can be referenced by many. 

So you could do the following: 

Add a `FHeightFieldShapeBarrier CreateHeightFieldShapeBarrier(agxCollide::HeightField* HeightField);`  (see e.g. `AGXBarrierFactories::CreateCapsuleShapeBarrier` for reference. 

Then, return that from within `FTerrainBarrier::GetHeightField`.